src/test/groovy/unit/org/pih/warehouse/api/ProductApiControllerSpec.groovy

```diff
@@ -0,0 +1,121 @@
+package unit.org.pih.warehouse.api
+
+import grails.orm.PagedResultList
+import grails.testing.gorm.DataTest
+import grails.testing.web.controllers.ControllerUnitTest
+import java.util.stream.IntStream
+import org.grails.plugins.testing.GrailsMockHttpServletResponse
+import org.grails.web.json.JSONElement
+import org.grails.web.json.JSONObject
+import org.hibernate.Criteria
+import org.json.JSONArray
+import spock.lang.Shared
+import spock.lang.Specification
+import spock.lang.Unroll
+
+import org.pih.warehouse.api.ProductApiController
+import org.pih.warehouse.core.Location
+import org.pih.warehouse.forecasting.ForecastingService
+import org.pih.warehouse.product.Product
+import org.pih.warehouse.product.ProductService
+
+@Unroll
+class ProductApiControllerSpec extends Specification implements DataTest, ControllerUnitTest<ProductApiController> {
+
+	@Shared private def forecastingServiceStub
+	@Shared private def productServiceStub
+
+    void setupSpec() {
+		mockDomains(Product, Location)
+    }
+
+	void setup() {
+		productServiceStub = Stub(ProductService)
+		forecastingServiceStub = Stub(ForecastingService)
+
+		controller.forecastingService = forecastingServiceStub
+		controller.productService = productServiceStub
+	}
+
+	void 'getDemand should return correctly'() {
+		given: 'the following db data'
+		Product product = new Product().save(validate: false)
+		Location location = new Location().save(validate: false)
+
+		and: 'the following params'
+		String productIdToLookFor = shouldFindProduct ? product.id : '-1'
+		String locationIdToLookFor = shouldFindLocation ? location.id : '-1'
+
+		params.id = productIdToLookFor
+		session.warehouse = [id: locationIdToLookFor]
+
+		and: 'the following mocks'
+		forecastingServiceStub.getDemand(_, _, _) >> [:]
+
+		when:
+		controller.demand()
+
+		then:
+		JSONObject json = getJsonObjectResponse(controller.response)
+		if (shouldFindLocation) json.data.location != null
+		else json.data.location == null
```

Comment:
as for if/else statements, we used to write them with the `{ }`, in this case it'd be:
```groovy
if (shouldFindLocation) {
  json.data.location != null
} else {
  json.data.location == null
}
```
btw I have a question what the `json.data.location != null` does, as it would return a boolean (it compares the `json.data.location` with the `null`)?

Reply:
yes that makes sense. I'll fix.

Actually you're right. Spock treats anything in the "then:" block that returns a boolean the same as if you prefixed it with "assert". If the condition returns false, it'd fail the test. BUT if the statement is nested in an if or else, then you need to explicitly provide the "assert" for it to trigger. I'll fix